src/oss-tests/ProjectManagerTests/ProjectManagerFactoryTests.cs

```diff
@@ -0,0 +1,113 @@
+using System.Net.Http;
+using VisualStudio.TestTools.UnitTesting;
+
+[TestClass]
+public class ProjectManagerFactoryTests
+{
+    private readonly Mock<IHttpClientFactory> _mockFactory = new();
+
+    private readonly List<Dictionary<string, ProjectManagerFactory.ConstructProjectManager>> _overrideConstructors;
+    
+    public ProjectManagerFactoryTests()
+    {
+        MockHttpMessageHandler mockHttp = new();
+
+        _mockFactory.Setup(_ => _.CreateClient(It.IsAny<string>())).Returns(mockHttp.ToHttpClient());
+        
+        _overrideConstructors = new List<Dictionary<string, ProjectManagerFactory.ConstructProjectManager>>
+        {
+            new()
+            {
+                {"test", directory => new NuGetProjectManager(_mockFactory.Object, directory) } // Create a test type with the NuGetProjectManager.
+            },
+            new()
+            {
+                {"nuget", directory => new NuGetProjectManager(_mockFactory.Object, directory) } // Create a new nuget type and override the NuGetProjectManager.
+            },
+            new()
+            {
+                {"nuget", directory => new NPMProjectManager(_mockFactory.Object, directory) } // Create a new nuget type, but use an NPMProjectManager instead.
+            },
+            new()
+            {
+                {"nuget", directory => new NuGetProjectManager(_mockFactory.Object, directory) }, // Create a new nuget type and override the NuGetProjectManager.
+                {"npm", directory => new NPMProjectManager(_mockFactory.Object, directory) }, // Create a new npm type and override the NPMProjectManager.
+                {"pypi", directory => new PyPIProjectManager(_mockFactory.Object, directory) } // Create a new pypi type and override the PyPIProjectManager.
+            },
+        };
+    }
+
+    [DataTestMethod]
+    [DataRow("pkg:nuget/newtonsoft.json", typeof(NuGetProjectManager))]
+    [DataRow("pkg:npm/foo", typeof(NPMProjectManager))]
+    [DataRow("pkg:pypi/plotly", typeof(PyPIProjectManager))]
+    public void FactorySucceeds(string purlString, Type expectedManager)
+    {
+        ProjectManagerFactory projectManagerFactory = new(_mockFactory.Object);
+
+        PackageURL packageUrl = new(purlString);
+
+        Assert.AreEqual(expectedManager, projectManagerFactory.CreateProjectManager(packageUrl)?.GetType());
+    }
+    
+    [DataTestMethod]
+    [DataRow(null)] // Add nothing, remove nothing.
+    [DataRow(0)] // Add a "test" type with the NuGetProjectManager, remove nothing.
+    [DataRow(1)] // Add a "nuget" type with the NuGetProjectManager, overriding the default, remove nothing.
+    [DataRow(2)] // Add a "nuget" type but use an NPMProjectManager, overriding the default, remove nothing.
+    [DataRow(3)] // Add "nuget", "npm" and "pypi" types, overriding the defaults, remove nothing.
+    [DataRow(null, "nuget")] // Add nothing, remove "nuget".
+    [DataRow(null, "nuget", "npm")] // Add nothing, remove "nuget" & "npm".
```

Comment:
These data rows are rather convoluted. I wonder if in this case, it would be easier/clearer to just write these as separate tests and not try to parameterize them?

(Personally, when I see comments like these, it makes me think how we might use one or more methods with appropriate names to reduce or eliminate the need for comments to explain what the code is doing.)

Reply:
Yeah definitely kind of a mess, just wanted to get some tests going. I will clean it up now.

Reply:
@pmalmsten cleaned it up a lot I think.